### PR TITLE
[rotation_distortion] Use "delayed swap" solution to reduce rotation distortion

### PR DESCRIPTION
**Pull request by hellohuanlin**
The "size interpolation" solution didn't go well (more context [here](https://github.com/flutter/engine/pull/40412#issuecomment-1485938933)). Then a new solution came to my mind, and I call it **"delayed swap"**:

In the originally behavior, we swap the width/height immediately before the rotation, resulting in roughly ~4x distortion in the beginning. With "delayed swap" solution, we **swap the width/height right in the middle of the rotation** (i.e. delay the swap for half of the transition duration).

This new "delayed swap" solution gives us the same benefit as the "snapshot" solution: 
- reducing ~4x distortion to ~2x
- most distorted frames occur in the middle of rotation when it's moving the fastest, making it hard to notice

And it fixes the drawback of "snapshot" solution: 
- it works well with dynamic content like animation or video
- it doesn't have a ~0.5 second penalty when taking the snapshot

Looks pretty good on flutter gallery: 

https://user-images.githubusercontent.com/41930132/228383137-7cd09982-89a9-4c83-bf55-9431de708278.mp4

*List which issues are fixed by this PR. You must list at least one issue.*

Fixes https://github.com/flutter/flutter/issues/16322

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by dnfield — COMMENTED** [review]
This LGTM but would appreciate a look from one of the other reviewers who is more famliar/active with Objective C :)

**Comment by dnfield** [comment]
I think this is an improvement but there's still a pretty big jump in the middle.

What does this look like on a larger screen like the iPad pro?

**Comment by cyanglaz** [comment]
A crazy idea: is it possible to substitute a screenshot at the middle to avoid the "jump"?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 1]
```diff
@@ -1328,15 +1363,27 @@ - (void)viewDidLayoutSubviews {
...
 }
 
-// Updates _viewportMetrics physical padding.
+// Set _viewportMetrics physical size.
+- (void)setViewportMetricsSize {
+  // TODO(hellohuanlin): Use [self mainScreenIfViewLoaded] instead of [UIScreen mainScreen].
+  // This requires adding the view to window during unit tests, which calls multiple engine calls
+  // that is hard to mock since they take/return structs. An alternative approach is to partial mock
+  // the FlutterViewController to make view controller life cycle methods no-op, and insert
+  // this mock into the responder chain.
```
[the remark this answers is not in the document]
Comment (hellohuanlin) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1377]: Nice! I did not think about just partial mocking `mainScreenIfViewLoaded`

**Review by jmagman — APPROVED** [review]

**Comment by jmagman** [comment]
I think you need to rebase onto #40732.

**Comment by jmagman** [comment]
@hellohuanlin I'm not sure why there's a `canvaskit_text_styles_text_style_half_leading_overrides_paragraph_style_half_leading` golden change on this PR.  You can file an issue but don't let it block you.  Either push an empty commit or manually merge.

**Comment by hellohuanlin** [comment]
@jmagman I'm curious to learn how did you navigate this website? I have logged in and didn't find anything.

**Comment by jmagman** [comment]
> @jmagman I'm curious to learn how did you navigate this website? I have logged in and didn't find anything.

I'm in no way the expert, but I checked "Digests: Positive" at the top:
Screenshot 2023-04-13 at 3 12 21 PM

I couldn't find a good description of how to use the Skia Gold client UI, just https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package%3Aflutter

**Comment by FaisalMohammadi** [comment]
i have this issue in my app right now in flutter 3.10.5 can anybody tell how exactly should i do to delay the swap

**Comment by FaisalMohammadi** [comment]
Ok perfect, you mean then in flutter 3.13, because now i have 3.10.5 and i see huge streching in my widgets.

**Comment by hellohuanlin** [comment]
@FaisalMohammadi this seems like frame drop issue. are you doing anything heavy? Can you try it on an older version before this is landed, so we can compare?

**Comment by FaisalMohammadi** [comment]
you mean in a older simulator? 
the widget tree for that specific widget looks like this 
widget tree


i am using offstage to show bottom navigation bar persistantly in every page so when i call my widget directly from main and not having bottom navigation bar and the other stuff as shown in the widget tree, then the result is a bit faster.

**Comment by hellohuanlin** [comment]
@FaisalMohammadi I meant older Flutter version when this PR is not landed.

**Pull request merged**